### DISCUSSION TITLE
How to delete all widgets inside custom widget and rerender it with new content?

### DISCUSSION THREAD

**Discussion by MohamedElgamal** [Textualize/textual#3230] [Q&A] [answered]
Want to delete all the content of TabbedContent and replace it with new one because the data of all tabs will change due to certain action will occur in another event. 
I tried as following coded two function the first is responsible to delete all the childrens of the TabbedContent its code as following:
<pre>
    def unmount_elements(self):
        self.remove_children()
</pre>

and the other function is responsible to add the content of tabs again, I tried as following: 
<pre>
   def compose_filtering_users_requests(self):
        states = self.state_dao.get_states()
        for index, state in enumerate(states):
            self.current_state = state
            tab_pane = self.mount(TabPane(state.get_state(), id=f"state{state.get_id()}"))
            tab_pane.mount(Label("Test"))
</pre>
this scenario fails because self.mount return AwaitMount but i need the widget itself to add new items to created widget.
and also tried:
<pre>
   def compose_filtering_users_requests(self):
       self.compose()
</pre>

both functions are triggered and called with another class, and it is simplified version from my code any ideas?

**Comment by TomJGooding** [comment, marked as answer]
Might the [clear_panes](https://textual.textualize.io/widgets/tabbed_content/#textual.widgets._tabbed_content.TabbedContent.clear_panes) and [add_pane](https://textual.textualize.io/widgets/tabbed_content/#textual.widgets._tabbed_content.TabbedContent.add_pane) methods for TabbedContent help?

**Comment by MohamedElgamal** [comment]
I think it will not work because i am trying to recreate elements like following:

<pre>
 def compose(self) -> ComposeResult:
        states = self.state_dao.get_states()
        with TabbedContent():
            for index, state in enumerate(states):
                self.current_state = state
                with TabPane(state.get_state(), id=f"state{state.get_id()}"):
                    with VerticalScroll(id=f"state{state.get_id()}"):
                        yield StateTabContent(self.current_logged_user, state, f"StateTabContent{state.get_id()}")
</pre>

this is the old elements that will be created again with new data

**Reply by davep** [reply]
Is this in reply to @TomJGooding's suggestion of using `clear_panes` and `add_pane`? Because if you want to clear the content of a `TabbedContent`, and then add fresh panes, they are the methods you would use.

When replying, it can be helpful to reply to the comment you're replying to, and to be very clear as to why you believe a suggested answer isn't going to work.

**Comment by MohamedElgamal** [comment]
if i used add_pane , I will face the same problem i am facing with my code add_pane return AwaitMount object not a reference to the tab itself which means i will not be able to add widget to tab itself

**Reply by davep** [reply]
When you say "add widget to tab", I take it you mean add a widget to the `TabPane`? The thing is, `add_pane`'s `pane` parameter is a `TabPane`. You'd create the `TabPane`, add the widgets, then add it to the `TabbedContent` with `add_pane`.

**Reply by MohamedElgamal** [reply]
I attempted to add a widget to a pane as follows: I created a Custom Tab Pane, which serves as the parent for a Tab Pane that includes all the widgets to be included in the pane. Then, using the add_pane method, I instantiated an instance of the Custom Tab Pane and added it to the pane. However, when running the program, it freezes when fire event that responsible to change the content of Tabbed Content
the following complete code scenario : 
<pre>
**class StatesTabCreator(Widget):
    def __init__(self, username):
        super().__init__()
        self.state_dao = connection_dao
        self.current_logged_user = username
        self.current_state = ""

    def compose(self) -> ComposeResult:
        states = self.state_dao.get_states()
        with TabbedContent():
            for index, state in enumerate(states):
                self.current_state = state
                with TabPane(state.get_state(), id=f"state{state.get_id()}"):
                    with VerticalScroll(id=f"state{state.get_id()}"):
                        yield StateTabContent(self.current_logged_user, state, f"StateTabContent{state.get_id()}")

    def unmount_elements(self):
        self.query_one(TabbedContent).clear_panes()

    def compose_filtering_users_requests(self):
        states = self.state_dao.get_states()
        tabbed_content: TabbedContent = self.query_one(TabbedContent)
        for index, state in enumerate(states):
            tabbed_content.add_pane(CustomTabPane(username=self.current_logged_user,
                                                  state=state,
                                                  title=state.get_state()))

class CustomTabPane(TabPane):
    def __init__(self, username, state, title):
        super().__init__(title, id=f"state{self.state.get_id()}")
        self.current_logged_user = username
        self.state = state

    def compose(self) -> ComposeResult:
        with VerticalScroll(id=f"state{self.state.get_id()}"):
            yield StateTabContent(self.current_logged_user, self.state, f"StateTabContent{self.state.get_id()}")
</pre>

**Reply by davep** [reply]
I sense it might make sense to start with a simple stand-alone example of what @TomJGooding suggested; and then work from there. So here's a stand-alone minimal example of `add_pane` and `clear_panes` adding and clearing the panes without any sort of hang:

```python
from textual.app import App, ComposeResult
from textual.widgets import TabbedContent, TabPane, Label, Input, Footer

class TestPane(TabPane):

    def compose(self) -> ComposeResult:
        for n in range(3):
            yield Label(f"Example input {n}")
            yield Input(placeholder=f"Example input {n}")

class DynamicTabbedContentApp(App[None]):

    BINDINGS = [
        ("a", "add", "Add a new test pane"),
        ("c", "clear", "Clear all the panes")
    ]

    def compose(self) -> ComposeResult:
        with TabbedContent():
            yield TestPane("First One")
        yield Footer()

    def action_add(self) -> None:
        self.query_one(TabbedContent).add_pane(TestPane("Another one"))

    def action_clear(self) -> None:
        self.query_one(TabbedContent).clear_panes()

if __name__ == "__main__":
    DynamicTabbedContentApp().run()
```

can you confirm that that works for you too?

**Reply by MohamedElgamal** [reply]
thanks , I found the problem, I was setting the wrong id in the custom Tab Pane
